cpp/src/unary/math_ops.cu

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "unary_ops.cuh"
+#include <cudf/legacy/unary.hpp>
+#include <cudf/legacy/copying.hpp>
+
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+
+#include <cmath>
+#include <algorithm>
+#include <type_traits>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+// trig functions
+
+struct DeviceSin {
+    template<typename T,
+             typename std::enable_if_t<!std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
+    __device__
+    T operator()(T data) {
+        return std::sin(data);
+    }
+
+    template <typename T,
+              typename std::enable_if_t<std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
```

Comment:
This special code path for `bool8` seems annoying to have to maintain. I think it would be unnecessary if `bool8` had an implicit conversion operator to arithmetic types.

Reply:
That was one of the first things that I tried (adding an implicit / explicit conversion to the `bool8` type) - however it didn't seem to work. I discussed with @trxcllnt the summary from out chat was:
> oh so looking back at the error you're getting... std::sin only has overloads for double, float, and long double? so IIUC the compiler wouldn't automatically call the uint8_t conversion operator.

Reply:
However - if I am missing something and there is a way to do this - I agree it would be much nicer. Having this duplicate code in each function object is very irritating.

Reply:
I think something like this should work:
```
template <typename T, std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
operator T ();
```

The compiler should select the conversion to `float/double`.

Reply:
I gave a couple variants of this a shot:
```
  template <typename T, std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
  CUDA_HOST_DEVICE_CALLABLE operator T() {
    return static_cast<T>(static_cast<bool>(this->value));
  }
```
but no dice. I had seen your post to @harrism earlier as well:
> Yeah, basically add what the old bool8 had: https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/utilities/legacy/wrapper_types.hpp#L107
and tried something similar. 

Did you have something different in mind for the body of the conversion operator? I am not a TMP expert so I might be missing something simple. I will take a stab at this tmo again

Reply:
That looks right to me. What didn't work?

Reply:
You get the same error as before:
```
/home/choekstra/ws3/cudf/cpp/src/unary/math_ops.cu(38): error: more than one instance of overloaded function "std::sin" matches the argument list:
            function "sin(double)"
            function "std::sin(float)"
            function "std::sin(long double)"
            argument types are: (cudf::experimental::bool8)
```

Reply:
How about factoring out the casting so that all operators can just call a single function rather than maintaining two copies of every operator? Something that calls the actual op with casts only for types that need it, like this. Could be part of a base class that all device_ops inherit.

```
template<typename T, typename Op, std::enable_if_t<!std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
__device__ T normalized_unary_op()(Op op, T data) { return Op(data); }

template<typename T, std::enable_if_t<std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
__device__ T normalized_unary_op()(T data) { return static_cast<T>(Op(static_cast<float>(data)); }
```

Reply:
This needs to be resolved in the `bool8` type and not externally in functions like this. I'm not sure what the right combinations of implicit/explicit conversion and operators is necessary, but it's not sustainable to have to do things like what was done here in providing two versions of every operator.